### ISSUE TITLE
Derive Vulkan bindings from D3D register syntax

### ISSUE THREAD

**Issue by tangent-vector** [shader-slang/slang#929]
Suppose a user writes code like:

```hlsl
SamplerState mySampler : register(s0, space2);
```

They might assume that when compiling for Vulkan, this will output as:

```glsl
layout(binding=0, set=2)
sampler mySampler;
```

We already emit a warning in this case (see issue #505) to indicate to user that they should use `[[vk::binding(0,2)]]` if they want to specify the layout for Vulkan, but it appears the wording of that message isn't clear enough. Some users are under the impression that the compiler is warning them, but then using their `register` specifier anyway.

From a user perspective, the feature request would be that when compiling for Vulkan/GLSL:

* If a parameter has a `[[vk::binding(...)]]`, then use that.

* Otherwise, if the parameter has a `register` binding, then attempt to use that instead.

This would need to work for both individual parameters and parameter blocks:

```hlsl
Texture2D t : register(t0, space3); // derive [[vk::binding(0,3)]]
ParameterBlock<X> pb : register(space4); // derive [[vk::binding(0,4)]]
```

There are some gotchas here that need to be thought through. The most obvious is when the simple derivation leads to overlapping bindings:

```hlsl
// This code is in error on Vulkan, but what should we report?
Texture2D t : register(t0, space2);
SamplerState s : register(s0, space2);
```

Diagnosing overlapping bindings here might be confusing to a user, since the `register` specifiers don't seem to overlap.

Another subtle issue is what to do when a parameter uses multiple resource kinds and thus needs multiple `register` specifiers:

```hlsl
struct MyStuff { Texture2D t; SamplerState s; };
MyStuff myStuff : register(t2) : register(s9);
```

In this case should we derive that `myStuff` gets a `binding` of `2` or of `9`? It seems that in a case like that we should diagnose an error, but that requires detecting the situation and explaining it to the user in reasonable terms.

I suppose another case down that road is when a `regsiter` specifier is present, but has nothing to do with the reosurces a parameter consumes:

```hlsl
Texture2D t : register(s0);
```

That should probably produce an error when compiling for D3D, but when compiling for Vulkan we don't have the information handy to know that the `s` register class is disjoint from our parameter (we just know that our parameter consumes `binding`s).

It is probably correct that all these edge cases are relatively rare, so that the payoff from having the feature outweighs the cost.

The other thing to consider if we go down this road is whether to support anything like the command-line options for dxc that apply and offset/shift to each of the D3D register classes when computing `binding`s. Those would be an additive change over the base feature, so we can consider them down the line.

**Comment by natduca** [comment]
Language cleanup, low, q3

**Comment by swoods-nv** [comment]
Fixed by #4338
